engine/src/flutter/impeller/geometry/matrix.cc

```diff
@@ -358,6 +360,75 @@ std::optional<MatrixDecomposition> Matrix::Decompose() const {
   return result;
 }
 
+std::optional<std::pair<Scalar, Scalar>> Matrix::GetScales2D() const {
+  if (HasPerspective2D()) {
+    return {};
+  }
+
+  // We only operate on the uppermost 2x2 matrix since those are the only
+  // values that can induce a scale on 2D coordinates.
+  // [ a b ]
+  // [ c d ]
+  double a = m[0];
+  double b = m[1];
+  double c = m[4];
+  double d = m[5];
+
+  if (b == 0.0f && c == 0.0f) {
+    return {{std::abs(a), std::abs(d)}};
+  }
+
+  if (a == 0.0f && d == 0.0f) {
+    return {{std::abs(b), std::abs(c)}};
+  }
+
+  // Compute eigenvalues for the matrix (transpose(A) * A):
+  //   [ a2  b2 ] == [ a  b ] [ a  c ] == [ aa + bb   ac + bd ]
+  //   [ c2  d2 ]    [ c  d ] [ b  d ]    [ ac + bd   cc + dd ]
+  // (note the reverse diagonal entries in the answer are identical)
+  double a2 = a * a + b * b;
+  double b2 = a * c + b * d;
+  double c2 = b2;
+  double d2 = c * c + d * d;
+
+  //
+  //   If L is an eigenvalue, then
+  //   det(this - L*Identity) == 0
+  //   det([ a - L     b   ]
+  //       [   c     d - L ]) == 0
+  //   (a - L) * (d - L) - bc == 0
+  //   ad - aL - dL + L^2 - bc == 0
+  //   L^2 + (-a + -d)L + ad - bc == 0
+  //
+  // Using quadratic equation for (Ax^2 + Bx + C):
+  //   A == 1
+  //   B == -(a2 + d2)
+  //   C == a2d2 - b2c2
+  //
+  // (We use -B for calculations because the square is the same as B and we
+  //  need -B for the final quadratic equation computations anyway.)
+  double negB = a2 + d2;
+  double C = a2 * d2 - b2 * c2;
+  double B2minus4AC = negB * negB - 4 * 1.0f * C;
```

Comment:
Here and elsewhere, use [`underscore_case`](https://google.github.io/styleguide/cppguide.html#Variable_Names)

Reply:
That gets tricky for these values that represent known variables in a common formula, so these would need to have capital A,B,C in them - but otherwise I can make them snake case.

Reply:
How about the new names?